### PR TITLE
Update NLPModelsJuMP for MathOptInterface

### PR DESCRIPTION
**Pull request by amontoison**
Interface between JuMP models (only composed of @NLobjective and @NLconstraint) and NLPModels. 
#22

### CODE REVIEW

File: docs/src/tutorial.md [hunk 1]
```diff
@@ -139,12 +139,12 @@ x, fx, ngx, optimal, iter = steepest(adnlp)
 The package
 [OptimizationProblems](https://github.com/JuliaSmoothOptimizers/OptimizationProblems.jl)
 provides a collection of problems defined in JuMP format, which can be converted
-to `MathProgNLPModel`.
+to `MathOptNLPModel`.
 
-```@example jumpnlp
+```julia
```
Comment (abelsiqueira) [inline docs/src/tutorial.md]: Does this work? `@example jumpnlp` was supposed to keep the previous environment.

File: src/moi_nlp_model.jl [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+end
+
+function NLPModels.jtprod!(nlp :: MathOptNLPModel, x :: AbstractVector, v :: AbstractVector, Jtv :: AbstractVector)
+  nlp.counters.neval_jac -= 1
+  increment!(nlp, :neval_jtprod)
+  Jtv .= jac(nlp, x)' * v
+  return Jtv
+end
+
+# Uncomment when :JacVec becomes available in MOI.
```
Comment (abelsiqueira) [inline src/moi_nlp_model.jl]: Is this still an issue?

Comment (amontoison) [reply src/moi_nlp_model.jl]: Yes, JacVec is still an unsupported feature...

File: docs/src/tutorial.md (hunk 1 above)
Comment (amontoison) [reply docs/src/tutorial.md]: Yes, but the code is not run. It's what I wanted because OptimizationProblems.jl is not working with JuMP 0.20.

I tested locally with an updated version of OptimizationProblems.jl and it works.
https://github.com/amontoison/OptimizationProblems.jl/commit/cf02b595bd0f865b954c42dad909490a85c9b967

**Comment by abelsiqueira** [comment]
I noticed you removed `NLS` models. Can they be remade?
Also, this needs the latest NLPModels, right?

File: docs/src/tutorial.md (hunk 1 above)
Comment (amontoison) [reply docs/src/tutorial.md]: I modified problems such that we have only @NLconstraint, @NLobjective and @NLexpression, it's why I didn't open a pull request.

Otherwise all models work without modifications on JuMP 0.20, we only need to replace `setvalue` by `set_start_value` to remove warnings.

**Comment by amontoison** [comment]
I think it can be remade but at which cost.... They completely changed how linear / quadratic terms are treated inside MOI. If we find a way to parse them, we should build ourself the jacobian , hessian...

With MPB everything was accessible through an MPB.NLPEvaluator (linear or non linear objective / constraints) and a quadratic term was a non linear term. With MOI they made a distinction between a quadratic term a non linear term. With MOI only restrictive non linear objective / constraints are accessible through MOI.NLPEvaluator.

I update the Project.toml to use the last release of NLPModels. I didn't tested with previous version.

**Comment by abelsiqueira** [comment]
You can't get the expressions anymore? MPBNLS was essentially creating constraints for an internal model using expressions.

**Comment by abelsiqueira** [comment]
> I update the Project.toml to use the last release of NLPModels. I didn't test with previous version.

I meant the `master` version, because `test/problems` is not in the latest released version.

**Comment by amontoison** [comment]
No It doesn't need the `master`version of NLPModels.
Problems are in `test/problems` folder for `NLPModelJuMP` and `test` folder for `NLPModel`.

File: src/moi_nlp_model.jl [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+  return c
+end
+
+function NLPModels.jac_structure!(nlp :: MathOptNLPModel, rows :: AbstractVector{<: Integer}, cols :: AbstractVector{<: Integer})
+  rows[1:nlp.meta.nnzj] .= nlp.jrows
+  cols[1:nlp.meta.nnzj] .= nlp.jcols
+  return (nlp.jrows, nlp.jcols)
+end
+
+function NLPModels.jac_structure(nlp :: MathOptNLPModel)
```
Comment (dpo) [inline src/moi_nlp_model.jl]: This is the default method, that is already implemented in NLPModels.jl.

File: src/moi_nlp_model.jl [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+end
+
+function NLPModels.jac_coord!(nlp :: MathOptNLPModel, x :: AbstractVector, vals :: AbstractVector)
+  increment!(nlp, :neval_jac)
+  MOI.eval_constraint_jacobian(nlp.eval, nlp.jvals, x)
+  vals[1:nlp.meta.nnzj] .= nlp.jvals
+  return vals
+end
+
+function NLPModels.jac(nlp :: MathOptNLPModel, x :: AbstractVector)
```
Comment (dpo) [inline src/moi_nlp_model.jl]: Already implemented.

File: src/moi_nlp_model.jl [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+function NLPModels.jtprod!(nlp :: MathOptNLPModel, x :: AbstractVector, v :: AbstractVector, Jtv :: AbstractVector)
+  nlp.counters.neval_jac -= 1
+  increment!(nlp, :neval_jtprod)
+  Jtv .= jac(nlp, x)' * v
+  return Jtv
+end
+
+# Uncomment when :JacVec becomes available in MOI.
+#
+# function NLPModels.jprod(nlp :: MathOptNLPModel, x :: AbstractVector, v :: AbstractVector)
```
Comment (dpo) [inline src/moi_nlp_model.jl]: Should disappear.

File: src/moi_nlp_model.jl [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+#   return jprod!(nlp, x, v, jv)
+# end
+
+# function NLPModels.jprod!(nlp :: MathOptNLPModel, x :: AbstractVector, v :: AbstractVector, jv :: AbstractVector)
+#   increment!(nlp, :neval_jprod)
+#   MOI.eval_constraint_jacobian_product(nlp.eval, jv, x, v)
+#   return jv
+# end
+
+# function NLPModels.jtprod(nlp :: MathOptNLPModel, x :: AbstractVector, v :: AbstractVector)
```
Comment (dpo) [inline src/moi_nlp_model.jl]: Should go.

File: src/moi_nlp_model.jl [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+# end
+
+function NLPModels.hess_structure!(nlp :: MathOptNLPModel, rows :: AbstractVector{<: Integer}, cols :: AbstractVector{<: Integer})
+  rows[1:nlp.meta.nnzh] .= nlp.hrows
+  cols[1:nlp.meta.nnzh] .= nlp.hcols
+  return (nlp.hrows, nlp.hcols)
+end
+
+function NLPModels.hess_structure(nlp :: MathOptNLPModel)
+  return (nlp.hrows, nlp.hcols)
```
Comment (dpo) [inline src/moi_nlp_model.jl]: Should go.

File: test/problems/hs10.jl [hunk 8]
```diff
@@ -4,8 +4,8 @@ function hs10()
   nlp = Model()
 
   @variable(nlp, x[i=1:2])
-  setvalue(x[1], -10)
-  setvalue(x[2],  10)
+  set_start_value(x[1], -10.0)
+  set_start_value(x[2],  10.0)
```
Comment (dpo) [inline test/problems/hs10.jl]: At other places, you left Ints as initial value. If they are accepted, I think 10 is clearer than 10.0.

**Comment by amontoison** [comment]
I didn't removed `jac`, `jac_structure` and `hess_structure` functions because their implementation for an `MathOptNLPModel` are more optimized than the generic ones for an `AbstractNLPModel`.

The main reason is that `jrows`, `jcols`, `hrows` and `hcols` are stored inside the model. `jac_structure`/`hess_structure` don't need to call  `jac_structure!`/`hess_structure!` and `jac` don't need to call `jac_structure`.

File: .travis.yml [hunk 9]
```diff
@@ -16,10 +16,6 @@ matrix:
   allow_failures:
     - julia: nightly
 
-addons:
-  apt_packages:
-    - gfortran # For Ipopt
-
```
Comment (dpo) [inline .travis.yml]: This shouldn't be part of this PR.

File: appveyor.yml [hunk 10]
```diff
@@ -2,6 +2,8 @@ environment:
   matrix:
   - julia_version: 1.0
   - julia_version: 1.1
+  - julia_version: 1.2
+  - julia_version: 1.3
```
Comment (dpo) [inline appveyor.yml]: This is already in #32.

File: .travis.yml (hunk 9 above)
Comment (amontoison) [reply .travis.yml]: I see that you just add it in #32. It should disappear after a rebase.

**Comment by amontoison** [comment]
I close this pull request, #38 includes `MathOptNLPModel` and `MathOptNLSModel`.
